src/OrleansDockerUtils/DockerMembershipOracle.cs

```diff
@@ -0,0 +1,468 @@
+                }
+
+                // Identify removed silos.
+                foreach (var silo in silos)
+                {
+                    // Never remove self.
+                    if (silo.Key.Equals(SiloAddress)) continue;
+
+                    // Do not remove silos which were present in the update.
+                    if (silo.Value.Refreshed)
+                    {
+                        // Reset the refresh flag.
+                        silo.Value.Refreshed = false;
+                        continue;
+                    }
+
+                    // The silo was not present in the update, remove it.
+                    if (removed == null) removed = new HashSet<SiloAddress>();
+                    removed.Add(silo.Key);
+                }
+
+                // If any silos were removed, 
+                if (removed != null)
+                {
+                    foreach (var silo in removed)
+                    {
+                        silos.Remove(silo);
+                    }
+                }
+            }
+
+            // If anything was updated, clear the cache before notifying clients.
+            if (added != null || removed != null)
+            {
+                // Clear the caches.
+                ClearCaches();
+
+                // If any silos were added or removed, notify subscribers of the new status.
+                var siloStatusListeners = subscribers.Values.ToList();
+                if (added != null)
+                {
+                    foreach (var address in added)
+                    {
+                        NotifySubscribers(address, SiloStatus.Active, siloStatusListeners);
+                    }
+                }
+
+                if (removed != null)
+                {
+                    foreach (var address in removed)
+                    {
+                        NotifySubscribers(address, SiloStatus.Dead, siloStatusListeners);
+                    }
+                }
+            }
+        }
+
+        private async Task RefreshAsync()
+        {
+            try
+            {
+                await resolver.Refresh();
+
+                lastRefreshTime = DateTime.UtcNow;
+            }
+            catch (Exception exception)
+            {
+                log?.Warn(
+                    (int)ErrorCode.Docker_MembershipOracle_ExceptionRefreshingSilos,
+                    "Exception refreshing silos.",
+                    exception);
+                throw;
+            }
+        }
+
+        /// <summary>
+        /// Clears the cached data.
+        /// </summary>
+        private void ClearCaches()
+        {
+            activeSilosCache = null;
+            allSilosCache = null;
+            multiClusterSilosCache = null;
+        }
+
+        /// <summary>
+        /// Updates the status of this silo.
+        /// </summary>
+        /// <param name="status">The updated status.</param>
+        private void UpdateStatus(SiloStatus status)
```

Comment:
The propagation of silo status changes happens in [FabricMembershipOracle.cs:L348](https://github.com/dotnet/orleans/pull/2542/files#diff-287349e1f70e4eabfc71ecef5e59735bR342). The only states which are supported are `Active` and `Dead`, but they satisfy the distinction for `SiloStatus.IsTerminating()`.

Do you think this is an issue, @gabikliot?